File: kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/KafkaTopicInfoSource.scala
/*
 * SPDX-FileCopyrightText: 2023 Marshall Wace <opensource@mwam.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.mwam.kafkakewl.metrics.services

import com.mwam.kafkakewl.metrics.domain.KafkaTopicPartitionInfoExtensions.*
import com.mwam.kafkakewl.metrics.domain.{KafkaTopicPartition, KafkaTopicPartitionInfo, KafkaTopicPartitionInfoChanges, KafkaTopicPartitionInfos}
import org.apache.kafka.common.TopicPartition
import zio.*
import zio.kafka.consumer.Consumer
import zio.stream.*

class KafkaTopicInfoSource(
  private val scope: Scope, // TODO This is slightly dangerous, the scope must not be used after it's closed
  private val topicInfoChangesStream: ZStream[Any, Throwable, KafkaTopicPartitionInfoChanges],
  private val hub: Hub[KafkaTopicPartitionInfoChanges]
) {
  def subscribe(): URIO[Scope, Dequeue[KafkaTopicPartitionInfoChanges]] = hub.subscribe
  def startPublishing(): UIO[Unit] = for {
    topicInfoChangesFiber <- topicInfoChangesStream.runScoped(ZSink.foreach(hub.publish)).provideEnvironment(ZEnvironment(scope)).orDie.fork // TODO error handling!
    // TODO not expecting subscribers after this -> those would not get the initial snapshot of topic infos, only diffs
    // TODO check if we need to interrupt the fiber at all, maybe runScoped does it already?
    _ <- scope.addFinalizer(topicInfoChangesFiber.interrupt *> hub.shutdown)
  } yield ()
}

object KafkaTopicInfoSource {
  def live: ZLayer[Consumer, Nothing, KafkaTopicInfoSource] =
    ZLayer.scoped {
      for {
        scope <- ZIO.service[Scope]
        consumer <- ZIO.service[Consumer]
        hub <- Hub.bounded[KafkaTopicPartitionInfoChanges](requestedCapacity = 1)
      } yield KafkaTopicInfoSource(scope, createKafkaTopicInfosStream(consumer), hub)
    }

  private def createKafkaTopicInfosStream(consumer: Consumer, timeout: Duration = 10.seconds): ZStream[Any, Throwable, KafkaTopicPartitionInfoChanges] = {
    val queryTopicPartitionInfos = (for {
      kafkaTopicPartitionInfos <- consumer.listTopics(timeout)
      topicPartitions = kafkaTopicPartitionInfos
        // TODO because otherwise the requests times out, probably there are some corrupt topics?
        .filter { case (topic, _) => topic.startsWith("kewltest") || topic.startsWith("streambook") || topic.startsWith("dbz") }
        .flatMap { case (topic, kafkaPartitionInfos) => kafkaPartitionInfos.map(tpi => new TopicPartition(topic, tpi.partition)) }
        .toSet
      beginningEndOffsets <- consumer.beginningOffsets(topicPartitions, timeout) zip consumer.endOffsets(topicPartitions, timeout)
    } yield {
      val (beginningOffsets, endOffsets) = beginningEndOffsets
      // TODO should indicate if there are topics for which we don't have either beginning or end-offsets or both
      (beginningOffsets.keySet intersect endOffsets.keySet)
        .view
        .map { tp =>
          (
            KafkaTopicPartition(tp),
            KafkaTopicPartitionInfo(beginningOffsets(tp), endOffsets(tp))
          )
        }
        .toMap
    }).tapError(t => ZIO.logError(s"failed to get the topics' beginning and end offsets: ${t.getMessage}"))
      .fold(_ => None, topicPartitionInfos => Some(topicPartitionInfos))

    ZStream.repeatZIOWithSchedule(queryTopicPartitionInfos.timed, Schedule.fixed(5.seconds))
      // removing the none items because they are failed queries
      .collect { case (duration, Some(topicPartitionInfos)) => (duration, topicPartitionInfos) }
      .tap { case (duration, topicPartitionInfos) => ZIO.logInfo(s"received ${topicPartitionInfos.size} topic-partition beginning-end offsets in ${duration.toMillis / 1000.0} seconds") }
      .map { case (_, topicPartitionInfos) => topicPartitionInfos }
      // putting the previous and current topic infos together, so that we can diff them below and publish the changes only
      .scan((KafkaTopicPartitionInfos.empty, KafkaTopicPartitionInfos.empty)) {
        case ((_, previous), current) => (previous, current)
      }
      // first emitted output of scan (2 empty maps) isn't needed
      .filter { case (previous, current) => previous.nonEmpty || current.nonEmpty }
      // diffing the current to the previous and emitting the changes
      .map { case (previous, current) => previous.diff(current) }
      .tap { changes => ZIO.logInfo(s"emitting ${changes.addedOrUpdated.size} added or updated topic-partition infos and ${changes.removed.size} removed ones")}
  }
}

File: kafkakewl-metrics/src/main/kotlin/com/mwam/kafkakewl/metrics/services/KafkaTopicInfoSource.kt
/*
 * SPDX-FileCopyrightText: 2024 Marshall Wace <opensource@mwam.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.mwam.kafkakewl.metrics.services

import arrow.core.Either
import arrow.core.raise.either
import com.mwam.kafkakewl.common.config.KafkaClientConfig
import com.mwam.kafkakewl.common.config.toKafkaClientConfig
import com.mwam.kafkakewl.metrics.TopicInfoSourceConfig
import com.mwam.kafkakewl.metrics.domain.KafkaTopicPartition
import com.mwam.kafkakewl.metrics.domain.KafkaTopicPartitionInfo
import com.mwam.kafkakewl.metrics.domain.KafkaTopicPartitionInfoChanges
import com.mwam.kafkakewl.metrics.domain.KafkaTopicPartitionInfos
import com.mwam.kafkakewl.metrics.domain.diff
import com.mwam.kafkakewl.utils.toEither
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.serialization.BytesDeserializer
import org.apache.kafka.common.utils.Bytes
import kotlin.time.Clock
import kotlin.time.ExperimentalTime
import kotlin.time.measureTimedValue
import kotlin.time.toJavaDuration

interface KafkaTopicInfoSource : AutoCloseable {
    fun topicInfoChanges(): SharedFlow<KafkaTopicPartitionInfoChanges>
    fun startPublishing(): Unit
}

private val logger = KotlinLogging.logger {}

class KafkaTopicInfoSourceImpl(
    kafkaClientConfig: KafkaClientConfig,
    private val config: TopicInfoSourceConfig,
) : KafkaTopicInfoSource {
    private val javaTimeout: java.time.Duration = config.timeout.toJavaDuration()
    private val excludeTopicRegexes: List<Regex> = config.excludeTopicRegexes.map { it.toRegex() }
    private val consumer: KafkaConsumer<Bytes, Bytes>
    private val sharedFlow = MutableSharedFlow<KafkaTopicPartitionInfoChanges>()
    private val scope = CoroutineScope(Dispatchers.Default)
    private var topicInfoPollingJob: Job? = null

    init {
        val consumerConfig = kafkaClientConfig.toKafkaClientConfig()
        consumer = KafkaConsumer<Bytes, Bytes>(consumerConfig, BytesDeserializer(), BytesDeserializer())
    }

    override fun topicInfoChanges(): SharedFlow<KafkaTopicPartitionInfoChanges> = sharedFlow

    override fun startPublishing() {
        topicInfoPollingJob = scope.launch {
            var lastTopicPartitionInfos: KafkaTopicPartitionInfos? = null
            while (isActive) {
                val (topicPartitionInfos, queryTopicPartitionInfosDuration) = measureTimedValue { queryTopicPartitionInfos() }

                topicPartitionInfos.fold(
                    { error ->
                        logger.info { "Polling topic infos finished in ${queryTopicPartitionInfosDuration.inWholeMilliseconds} ms, but $error" }
                    },
                    { topicPartitionInfos ->
                        val topicPartitionInfoChanges = if (lastTopicPartitionInfos != null) {
                            lastTopicPartitionInfos.diff(topicPartitionInfos)
                        } else {
                            KafkaTopicPartitionInfoChanges(topicPartitionInfos, emptySet())
                        }
                        lastTopicPartitionInfos = topicPartitionInfos

                        logger.info { "Polling topic infos finished in ${queryTopicPartitionInfosDuration.inWholeMilliseconds} ms, returned ${topicPartitionInfos.size} topic-partition-infos: $topicPartitionInfoChanges" }
                        if (!topicPartitionInfoChanges.isEmpty) {
                            sharedFlow.emit(topicPartitionInfoChanges)
                        }
                    }
                )

                delay(config.pollInterval)
            }
        }
    }

    override fun close() {
        logger.info { "Closing..." }
        runBlocking {
            scope.cancel()
            topicInfoPollingJob?.join()
        }
        consumer.close()
        logger.info { "Finished closing" }
    }

    private fun queryTopicPartitionInfos(): Either<String, KafkaTopicPartitionInfos> = either {
        val offsetTimestamp = Clock.System.now()
        val topicPartitionInfos = consumer.runCatching { listTopics(javaTimeout) }.toEither { "listTopics() failed: ${it.message}" }.bind()

        // filtering the topics' partition infos to the ones that we should not exclude
        val filteredTopicPartitionInfos = if (excludeTopicRegexes.isNotEmpty()) {
            topicPartitionInfos.filter { topicWithPartitionInfos -> excludeTopicRegexes.none { it.matches(topicWithPartitionInfos.key!!) } }
        } else {
            topicPartitionInfos
        }

        val topicPartitions = filteredTopicPartitionInfos.values.flatMap { it.map { TopicPartition(it.topic(), it.partition()) } }

        val beginningOffsets = consumer.runCatching { consumer.beginningOffsets(topicPartitions, javaTimeout) }.toEither { "beginningOffsets(${topicPartitions.size} topic-partitions) failed: ${it.message}" } .bind()
        val endOffsets = consumer.runCatching { consumer.endOffsets(topicPartitions, javaTimeout) }.toEither { "endOffsets(${topicPartitions.size} topic-partitions) failed: ${it.message}" } .bind()

        // Logging the missing beginning / end offsets
        fun Set<TopicPartition?>.logMissingOffsets(offsetType: String): Unit {
            if (this.isNotEmpty()) {
                logger.warn { "Could not get $offsetType-offsets for topic-partitions: $this" }
            }
        }
        val topicPartitionSet = topicPartitions.toSet()
        topicPartitionSet.subtract(beginningOffsets.keys).logMissingOffsets("beginning")
        topicPartitionSet.subtract(endOffsets.keys).logMissingOffsets("end")

        // Finally returning the ones with both beginning and end offsets
        beginningOffsets.keys.intersect(endOffsets.keys)
            .associate { KafkaTopicPartition(it) to KafkaTopicPartitionInfo(beginningOffsets[it]!!, endOffsets[it]!!, offsetTimestamp) }
    }
}
